pkg/controllers/management/etcdbackup/etcdbackup.go

```diff
@@ -0,0 +1,277 @@
+	if err != nil {
+		return err
+	}
+
+	newestBackup := &v3.EtcdBackup{}
+	for _, clusterBackup := range clusterBackups {
+		if len(newestBackup.Name) == 0 {
+			newestBackup = clusterBackup
+			continue
+		}
+		if getBackupCompletedTime(clusterBackup).After(getBackupCompletedTime(newestBackup)) {
+			newestBackup = clusterBackup
+		}
+	}
+	// this cluster has backups, lets see if the last one is old enough
+	// a new backup is due if this is true
+	if time.Since(getBackupCompletedTime(newestBackup)) > backupIntervalHours {
+		newBackup := NewBackupObject(cluster)
+		v3.BackupConditionCreated.CreateUnknownIfNotExists(newBackup)
+		newBackup, err := c.backupClient.Create(newBackup)
+		if err != nil {
+			logrus.Infof("[etcd-backup] Backup create failed:%v", err)
+			return err
+		}
+		logrus.Infof("[etcd-backup] New backup created: %s", newBackup.Name)
+	}
+
+	// rotate old backups
+	retention := cluster.Spec.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig.Retention
+
+	expiredBackups := getExpiredBackups(retention, internvalHours, clusterBackups)
+	for _, backup := range expiredBackups {
+		if err := c.backupClient.DeleteNamespaced(backup.Namespace, backup.Name, &metav1.DeleteOptions{}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func NewBackupObject(cluster *v3.Cluster) *v3.EtcdBackup {
+	controller := true
+	return &v3.EtcdBackup{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    cluster.Name,
+			GenerateName: fmt.Sprintf("%s-", cluster.Name),
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					Name:       cluster.Name,
+					UID:        cluster.UID,
+					APIVersion: cluster.APIVersion,
+					Kind:       cluster.Kind,
+					Controller: &controller,
+				},
+			},
+		},
+		ClusterID: cluster.Name,
+	}
+}
+
+func getBackupFilename(snapshotName string, cluster *v3.Cluster) string {
+	if cluster.Spec.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig == nil ||
+		cluster.Spec.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig.S3BackupConfig == nil {
+		return ""
+	}
+	target := cluster.Spec.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig.S3BackupConfig
+	return fmt.Sprintf("https://s3-%s.amazonaws.com/%s/%s", target.Region, target.BucketName, snapshotName)
+}
+
+func deleteS3Snapshot(b *v3.EtcdBackup) error {
+	if b.BackupConfig.S3BackupConfig == nil {
+		return fmt.Errorf("can't find S3 backup target configuration")
+	}
+	var err error
+	s3Client := &minio.Client{}
+	endpoint := b.BackupConfig.S3BackupConfig.Endpoint
+	bucket := b.BackupConfig.S3BackupConfig.BucketName
+	// no access credentials, we assume IAM roles
+	if b.BackupConfig.S3BackupConfig.AccessKey == "" ||
+		b.BackupConfig.S3BackupConfig.SecretKey == "" {
+		creds := credentials.NewIAM("")
+		s3Client, err = minio.NewWithCredentials(endpoint, creds, true, "")
+	} else {
+		accessKey := b.BackupConfig.S3BackupConfig.AccessKey
+		secretKey := b.BackupConfig.S3BackupConfig.SecretKey
```

Comment:
with latest changes, secretKey would need to be dereferenced, adding `management.Core.Secrets` to Controller and 
```
value, err := password.GetValueForPasswordField(b.BackupConfig.S3BackupConfig.SecretKey, c.secrets)
```

Reply:
forgot mentioning, password here is 
 `password "github.com/rancher/rancher/pkg/api/store/password"`

Reply:
This will not work due to how kontainer-engine passes the spec.